src/logging.cpp

```diff
@@ -222,23 +226,43 @@ void BCLog::Logger::LogPrintStr(const std::string &str)
     if (m_print_to_file) {
         std::lock_guard<std::mutex> scoped_lock(m_file_mutex);
 
+        // While the system is initializing, save some reasonable
+        // number of log messages for re-logging when the log
+        // file rotates (gets reopened).
+        if (m_relog_enable && m_msgs_relog.size() < 1000) {
+            m_msgs_relog.push_back(str_prefixed);
+        }
+
         // buffer if we haven't opened the log yet
         if (m_fileout == nullptr) {
             m_msgs_before_open.push_back(str_prefixed);
         }
         else
         {
+            FileWriteStr(str_prefixed, m_fileout);
+
             // reopen the log file, if requested
-            if (m_reopen_file) {
+            if (m_reopen_file && m_started_new_line) {
                 m_reopen_file = false;
                 FILE* new_fileout = fsbridge::fopen(m_file_path, "a");
                 if (new_fileout) {
                     setbuf(new_fileout, nullptr); // unbuffered
                     fclose(m_fileout);
                     m_fileout = new_fileout;
                 }
+                if (!m_relog_enable) {
+                    // rewrite the initial (configuration) messages
+                    FileWriteStr(
+                        Prefix("Logging: start relogging initial messages\n"),
+                        m_fileout);
+                    for (const std::string& msg : m_msgs_relog) {
+                        FileWriteStr(msg, m_fileout);
+                    }
+                    FileWriteStr(
+                        Prefix("Logging: end relogging initial messages\n"),
+                        m_fileout);
+                }
             }
-            FileWriteStr(str_prefixed, m_fileout);
```

Comment:
This line had to move up to before the file re-open, to fix the bug explained above.